include/range/v3/view/drop_last.hpp

```diff
@@ -0,0 +1,159 @@
+//
+
+#ifndef RANGES_V3_VIEW_DROP_LAST_HPP
+#define RANGES_V3_VIEW_DROP_LAST_HPP
+
+#include <type_traits>
+#include <meta/meta.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/view.hpp>
+#include <range/v3/view_interface.hpp>
+#include <range/v3/view_adaptor.hpp>
+#include <range/v3/utility/functional.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        template<typename Rng>
+        struct drop_last_bidirectional_view
+          : view_interface<drop_last_bidirectional_view<Rng>, range_cardinality<Rng>::value>
+        {
+            CONCEPT_ASSERT(ranges::BidirectionalRange<Rng>());
+        private:
+            friend range_access;
+            Rng rng;
+            std::size_t n;
+            detail::non_propagating_cache<sentinel_t<Rng>> sentinel;
+            void update_sentinel()
+            {
+                if (!sentinel)
+                {
+                    sentinel = ranges::prev(ranges::end(rng), n, ranges::begin(rng));
+                }
+            }
+        public:
+            drop_last_bidirectional_view() = default;
+            drop_last_bidirectional_view(Rng&& rng, std::size_t n)
+              : rng(std::move(rng)), n(n)
+            {}
+
+            iterator_t<Rng> begin()
+            {
+                return ranges::begin(rng);
+            }
+            sentinel_t<Rng> end()
+            {
+                update_sentinel();
+                return *sentinel;
+            }
+            Rng & base()
+            {
+                return rng;
+            }
+            Rng const & base() const
+            {
+                return rng;
+            }
+        };
+
+        template<typename Rng>
+        struct drop_last_forward_view
+          : view_adaptor<drop_last_forward_view<Rng>, Rng>
+        {
+            CONCEPT_ASSERT(ranges::ForwardRange<Rng>());
+        private:
+            friend range_access;
+            std::size_t n;
+
+            struct sentinel_adaptor;
+            class adaptor : public adaptor_base
+            {
+                friend sentinel_adaptor;
+                iterator_t<Rng> probe;
+            public:
+                adaptor() = default;
+                adaptor(const Rng& rng, std::size_t n)
+                {
+                    probe = ranges::next(ranges::begin(rng), n, ranges::end(rng));
+                }
+
+                void next(iterator_t<Rng>& it)
+                {
+                    ++it;
+                    ++probe;
+                }
+            };
+
+            struct sentinel_adaptor :  adaptor_base
+            {
+                bool empty(const iterator_t<Rng>&, const adaptor& ia, const sentinel_t<Rng>& s) const
+                {
+                    return ia.probe == s;
+                }
+            };
+
+            adaptor begin_adaptor() const { return {this->base(), n}; }
+            sentinel_adaptor end_adaptor() const { return {}; }
```

Comment:
Some views are not `const`-iterable (e.g. `view::filter`). If you want to be able to `drop_last` such a view, you need non-`const`-qualified `(begin|end)_adaptor` members.

Reply:
I looked at `view::remove_if` and `drop_while` implementations. It seems that all iterators/adaptors that have "underlying cache" does not have const version. Should I follow that?

That seems weird to me. That's interior mutability. Unrelated to data in range. Caching iterator does not mutate underlying value in any way.